src/main/kotlin/org/rust/cargo/toolchain/Cargo.kt

```diff
@@ -91,13 +91,8 @@ class Cargo(private val cargoExecutable: Path) {
     }
 
     @Throws(ExecutionException::class)
-    fun checkProject(
-        project: Project,
-        owner: Disposable,
-        projectDirectory: Path,
-        cargoPackage: CargoWorkspace.Package
-    ): ProcessOutput {
-        val arguments = mutableListOf("--message-format=json", "--package", cargoPackage.name)
+    fun checkProject(project: Project, cargoProjectDirectory: Path): ProcessOutput {
+        val arguments = mutableListOf("--message-format=json", "--all")
```

Comment:
@Undin but disposing of a platform module should not affect cargo projects anyhow?

Reply:
@vlad20012 And probably it's incorrect behavior and ideally we want to synchronize module and cargo project structure (at least in IDEA)
Because it possible to have the following sutuations:
1. User attaches new rust module to project via `Import module` action (i.e. new attached module can be located anywhere). New module will be shown in project structure but not in cargo toolwindow as separated cargo project
2. User attaches new cargo project via `Attach Cargo project` action (sources of this cargp project can be located anywhere). New cargo project will be shown in cargo toolwindow but not in project structure

Reply:
1.5 years ago the project structure was mapped to IDEA modules, and there were some problems (I didn't remember them). Then @matklad rewrote it.

Reply:
The fundamental problem is that modules are IDEA specific. There are modules in PyCharm, but they are hidden from UI, and PyCharm just creates a single dummy module for the whole project. 

The philosophical problem is that it's infeasible to maintain a losseless mapping between the truth (cargo project model) and the IDEA modules structure, for two reasons:
* they are just different structures. Rust models is build around the DAG of crates, IDEA's model is build around classpath
* user can manually modify the module structure, thus forcing inconsistencies between IDEA's and Cargo's model 

The practical problem is that `ExternalModel` subsystem is hellishly complicated (it has to be, because it wants to communicate with a live running gradle daemon via a binary interface) :) 

I feel like there's no ideal solution here. Is there some solution in the platform to show "fake" nodes in the project structure view? That is, a hook to show an entry in the UI tree of modules, **without** creating the corresponding `Module` object? We use a similar strategy for libraries (`SyntheticLibraryProvider` or something), and it works great: UI is what the user expects, but the underlying model is 100% Cargo based.

Reply:
>1.5 years ago

I'd say more like 2.5 years ago :)

Reply:
> PyCharm just creates a single dummy module for the whole project.

Now it's not true. You can attach the additional project to the existing one and PyCharm will create a separate module for new sources (and even allows having separate interpreters for each module)

Reply:
Yeah, that's the point: in IDEA, there's a "project structure" dialog where the user sees the list of modules and can modify them arbitrary.

PyCharm does not talk about "module"s , it has this "open two projects in a single window" action which works completely different to IDEA's UI. RubyMine, PhpStorm and CLion probably have their own unique hacks....

Reply:
In general, I don't think that we should implement synchronization between cargo projects and modules right now. My point is not to drop existing code that correctly works with the arbitrary disposable owner while running `cargo check`

Reply:
>My point is just not to drop existing code that correctly works with the arbitrary disposable owner while running `cargo check`

Oh, agree here! I haven't looked at the actual diff, was only responding to the comment :)